lib/Onyx.js

```diff
@@ -908,67 +965,46 @@ function multiSet(data) {
- * @param {String} key
- * @param {*} data
- *
+ * @param {*} existingValue
+ * @param {*} changes
  * @returns {*}
  */
-function applyMerge(key, data) {
-    const mergeValues = mergeQueue[key];
-    if (_.isArray(data) || _.every(mergeValues, _.isArray)) {
-        // Array values will always just concatenate
-        // more items onto the end of the array
-        return _.reduce(mergeValues, (modifiedData, mergeValue) => [
-            ...modifiedData,
-            ...mergeValue,
-        ], data || []);
+function applyMerge(existingValue, changes) {
+    if (_.isUndefined(existingValue) || _.isNull(existingValue)) {
+        return changes;
     }
 
-    if (_.isObject(data) || _.every(mergeValues, _.isObject)) {
-        // Object values are merged one after the other
-        return _.reduce(mergeValues, (modifiedData, mergeValue) => {
-            // lodash adds a small overhead so we don't use it here
-            // eslint-disable-next-line prefer-object-spread, rulesdir/prefer-underscore-method
-            const newData = Object.assign({}, fastMerge(modifiedData, mergeValue));
-
-            // We will also delete any object keys that are undefined or null.
-            // Deleting keys is not supported by AsyncStorage so we do it this way.
-            // Remove all first level keys that are explicitly set to null.
-            return _.omit(newData, (value, finalObjectKey) => _.isNull(mergeValue[finalObjectKey]));
-        }, data || {});
+    if (_.isArray(existingValue) || _.isArray(changes)) {
+        return changes;
+    }
+
+    if (_.isObject(existingValue) || _.isObject(changes)) {
+        // lodash adds a small overhead so we don't use it here
+        // eslint-disable-next-line prefer-object-spread, rulesdir/prefer-underscore-method
+        const newData = fastMerge(existingValue, changes);
+
+        // We will also delete any object keys that are undefined or null.
+        // Remove all first level keys that are explicitly set to null.
+        return _.omit(newData, value => _.isNull(value));
     }
 
     // If we have anything else we can't merge it so we'll
     // simply return the last value that was queued
-    return _.last(mergeValues);
+    return changes;
 }
 
 /**
  * Merge a new value into an existing value at a key.
  *
- * The types of values that can be merged are `Object` and `Array`. To set another type of value use `Onyx.set()`. Merge
- * behavior uses lodash/merge under the hood for `Object` and simple concatenation for `Array`. However, it's important
- * to note that if you have an array value property on an `Object` that the default behavior of lodash/merge is not to
- * concatenate. See here: https://github.com/lodash/lodash/issues/2872
+ * The types of values that can be merged are `Object` and `Array`. To set another type of value use `Onyx.set()`.
+ * Values of type `Object` get merged with the old value, whilst for `Array`'s we simply replace the current value with the new one.
```

Comment:
We did that, because there were no cases in the app, were we actually rely on this kind of merging. Also it doesn't really make sense to do that anyway.

Reply:
A different scenario that @perunt recently came up with was if we should check for top-level arrays and see if we can merge the contained items. For example we could replace onyx entries like `reportActions_<id>` with just `reportActions`

Reply:
Thanks for explaining why we decided to change the array merge behavior.

> A different scenario that @perunt recently came up with was if we should check for top-level arrays and see if we can merge the contained items.

So to provide an example it would be something like this right?
```javascript
const reportActions = [
    {
        id: 1,
        message: 'Hi',
        accountID: 1,
    },
    {
        id: 2,
        message: 'How are you?',
        accountID: 2,
    },
    {
        id: 1,
        message: 'Hi',
        accountID: 1,
    },
];

Ony.merge('reportActionsList', reportActions);

const newReportActions = [
    {
        id: 1,
        message: 'Hi',
        accountID: 1,
    },
    {
        id: 2,
        message: 'How are you?',
        accountID: 2,
    },
    {
        id: 3,
        message: 'I'm good thanks',
        accountID: 1,
    },
];

Ony.merge('reportActionsList', newReportActions);
```
The resulting reportActionsList in Onyx would contain exactly what we see in newReportActions, but it would only add the third object to the list since the other two already exist? Also, if we updated a field of an existing object then it would get merged in. If that's what you mean then I think that would be a really nice way for it to work.

Reply:
Yes, that's correct! Onyx should then not only compare top-level - in this case an array - but also every items by an identified by some primary key field.

P.S. i think the second item with `id: 1` in `reportActions` is a not intended, is it?

Reply:
>We did that, because there were no cases in the app, were we actually rely on this kind of merging. Also it doesn't really make sense to do that anyway.

We also do it this way because that's how SQLite MergePatch algorithm works by default and the array merging behavior we had was totally confusing and barely used.